packages/chain-bridge/src/adapters/StargateAdapter.sol

```diff
@@ -221,25 +250,62 @@ contract StargateAdapter is
                 stargateContract
             );
 
-        // Get messaging fee and perform transfer
-        MessagingFee memory messagingFee = stargate.quoteSend(sendParam, false);
-
-        if (providedFee < messagingFee.nativeFee) {
-            revert InsufficientFee(messagingFee.nativeFee, providedFee);
-        }
-
-        // Use exact fee amount from quote - Stargate handles refunds to keeper
-        stargate.sendToken{value: messagingFee.nativeFee}(
+        // Determine fee payment mode and get messaging fee
+        bool payInToken = options.payInProtocolToken &&
+            protocolFeeToken != address(0);
+        MessagingFee memory messagingFee = stargate.quoteSend(
             sendParam,
-            messagingFee,
-            params.refundAddress // Always refund to keeper who paid fees
+            payInToken
         );
 
-        // Refund any unused native value (buffer) back to the designated refund address
-        uint256 refundAmount = providedFee - messagingFee.nativeFee;
-        if (refundAmount > 0) {
-            (bool ok, ) = params.refundAddress.call{value: refundAmount}("");
-            if (!ok) revert RefundFailed(params.refundAddress, refundAmount);
+        if (payInToken) {
```

Comment:
Preferably move the code inside the `if` and the code inside the `else` to functions for easier read